packages/next/src/build/index.ts

```diff
@@ -2809,172 +2829,236 @@ export default async function build(
+              isRoutePPREnabled === true &&
+              config.experimental.pprFallbacks === true
 
-                    // normalize header values as initialHeaders
-                    // must be Record<string, string>
-                    for (const key of headerKeys) {
-                      // set-cookie is already handled - the middleware cookie setting case
-                      // isn't needed for the prerender manifest since it can't read cookies
-                      if (key === 'x-middleware-set-cookie') continue
+            if (!hasRevalidateZero && isDynamicRoute(originalAppPath)) {
+              // When PPR fallbacks aren't used, we need to include it here. If
+              // they are enabled, then it'll already be included in the
+              // prerendered routes.
+              if (!shouldUsePPRFallback) {
+                dynamicRoutes.push(page)
+              }
 
-                      let value = exportHeaders[key]
+              for (const route of dynamicRoutes) {
+                const normalizedRoute = normalizePagePath(route)
 
-                      if (Array.isArray(value)) {
-                        if (key === 'set-cookie') {
-                          value = value.join(',')
-                        } else {
-                          value = value[value.length - 1]
-                        }
-                      }
+                const { metadata, revalidate } =
+                  exportResult.byPath.get(route) ?? {}
 
-                      if (typeof value === 'string') {
-                        routeMeta.initialHeaders[key] = value
-                      }
-                    }
-                  }
+                let dataRoute: string | null = null
+                if (!isRouteHandler) {
+                  dataRoute = path.posix.join(`${normalizedRoute}${RSC_SUFFIX}`)
+                }
 
-                  prerenderManifest.routes[route] = {
-                    ...routeMeta,
-                    experimentalPPR,
-                    experimentalBypassFor: bypassFor,
-                    initialRevalidateSeconds: revalidate,
-                    srcRoute: page,
-                    dataRoute,
-                    prefetchDataRoute,
-                  }
-                } else {
-                  hasDynamicData = true
-                  // we might have determined during prerendering that this page
-                  // used dynamic data
-                  pageInfos.set(route, {
-                    ...(pageInfos.get(route) as PageInfo),
-                    isSSG: false,
-                    isStatic: false,
-                  })
+                let prefetchDataRoute: string | undefined
+
+                // While we may only write the `.rsc` when the route does not
+                // have PPR enabled, we still want to generate the route when
+                // deployed so it doesn't 404. If the app has PPR enabled, we
+                // should add this key.
```

Comment:
I think this is technically unchanged in your PR but I'm realizing I don't really understand what this comment is saying. The code below it is specifying the path to the `.prefetch.rsc` route when PPR is enabled and isn't doing anything related to output generation or `.rsc` right?

Reply:
I'll delete the comment as it also doesn't make sense 😅